// Code generated by github.com/tizz98/results; DO NOT EDIT.
package results

import (
	"context"
	"fmt"
)

type Int8Result struct {
	value *int8
	err   error
}

// SetNewInt8Result is a shortcut to creating a new Int8Result and then calling .Set(v, err) on it.
func SetNewInt8Result(v int8, err error) (result Int8Result) {
	result.Set(v, err)
	return
}

// SetNewInt8ResultPtr is a shortcut to creating a new Int8Result and then calling .Set(v, err) on it.
// This function differs from SetNewInt8Result by returning a pointer to Int8Result.
func SetNewInt8ResultPtr(v int8, err error) *Int8Result {
	result := SetNewInt8Result(v, err)
	return &result
}

// NewOptionalInt8Result is a shortcut to creating a new Int8Result and then calling .SetOptional(v, err) on it.
func NewOptionalInt8Result(v *int8, err error) (result Int8Result) {
	result.SetOptional(v, err)
	return
}

// NewOptionalInt8ResultPtr is a shortcut to creating a new Int8Result and then calling .SetOptional(v, err) on it.
// This function differs from NewOptionalInt8Result by returning a pointer to Int8Result.
func NewOptionalInt8ResultPtr(v *int8, err error) *Int8Result {
	result := NewOptionalInt8Result(v, err)
	return &result
}

// IsOk returns true when the result contains a non-nil result with no error
func (r Int8Result) IsOk() bool {
	return r.err == nil
}

// IsErr returns true when the result contains a non-nil error
func (r Int8Result) IsErr() bool {
	return r.err != nil
}

// Unwrap panics if the result contains an error, otherwise it returns the value
func (r Int8Result) Unwrap() int8 {
	if r.IsErr() {
		panic("cannot unwrap Int8Result, it is an error")
	}
	return *r.value
}

// UnwrapTo will call the .Err() method on the other Result if this Int8Result has an error.
// If other is a pointer to a Int8Result, then .Ok() will be called if this Int8Result name does not have an error.
func (r Int8Result) UnwrapTo(other Result) Result {
	if r.IsErr() {
		other.Err(r.GetErr())
	} else if other, ok := other.(*Int8Result); ok {
		other.Ok(r.Unwrap())
	}

	return other
}

// Expect panics with the specified message if the result contains an error, otherwise it returns the value
func (r Int8Result) Expect(message string) int8 {
	if r.IsErr() {
		panic(fmt.Errorf("%s: %w", message, r.GetErr()))
	}
	return *r.value
}

// Expectf panics with the specified message if the result contains an error, otherwise it returns the value.
// This is different than Expect because if will automatically format the string with the given args.
func (r Int8Result) Expectf(format string, args ...interface{}) int8 {
	if r.IsErr() {
		panic(fmt.Errorf("%s: %w", fmt.Sprintf(format, args...), r.GetErr()))
	}
	return *r.value
}

// UnwrapOr returns the value if there is not an error, otherwise the specified value is returned
func (r Int8Result) UnwrapOr(v int8) int8 {
	if r.IsOk() {
		return r.Unwrap()
	}
	return v
}

// UnwrapOrElse returns the value if there is not an error, otherwise the function is called and the result is returned
func (r Int8Result) UnwrapOrElse(fn func(err error) int8) int8 {
	if r.IsOk() {
		return r.Unwrap()
	}
	return fn(r.err)
}

// Ok sets the result to a successful result with the provided value.
// This will panic if the result has already been set to successful or an error.
func (r *Int8Result) Ok(v int8) {
	r.clear()
	r.value = &v
}

// Err sets the result to an error result with the provided error.
// This will panic if the result has already been set to successful or an error.
func (r *Int8Result) Err(err error) {
	r.clear()
	r.err = err
}

// GetError returns the error of the result. It may be nil, so check with Int8Result.IsErr() first.
func (r Int8Result) GetErr() error {
	return r.err
}

// Tup returns a tuple of (int8, error) with 0 being returned for int8 if there is an error
func (r Int8Result) Tup() (int8, error) {
	return r.UnwrapOr(0), r.err
}

// Set is a shortcut to checking the value of an error before setting the result.
// If there is an error, Int8Result.Err(err) will be called, otherwise Int8Result.Ok(v) will be called.
func (r *Int8Result) Set(v int8, err error) {
	if err != nil {
		r.Err(err)
	} else {
		r.Ok(v)
	}
}

// SetOptional is similar to Set but can be called with a nil value for int8.
// The error will be checked first to see if it is not nil, then if v is not nil it will be set with .Ok(v).
func (r *Int8Result) SetOptional(v *int8, err error) {
	if err != nil {
		r.Err(err)
	} else if v != nil {
		r.Ok(*v)
	}
}

func (r Int8Result) clear() {
	r.value = nil
	r.err = nil
}

// ResultToInt8Result takes a Result interface and returns a Int8Result. If "r" contains a Int8Result,
// it is returned, otherwise a new Int8Result is returned with an error set.
func ResultToInt8Result(r Result) (result Int8Result) {
	v, ok := r.(*Int8Result)
	if !ok {
		result.Err(fmt.Errorf("expected *Int8Result got %T", v))
		return
	}

	return *v
}

// ContextWithInt8 embeds the given value of int8 into the context for later retrieval with Int8FromContext
func ContextWithInt8(ctx context.Context, key interface{}, v int8) context.Context {
	return context.WithValue(ctx, key, v)
}

// Int8FromContext attempts to retrieve a int8 value from the specified context. A Int8Result is returned
// which can be used to inspect the success or failure of retrieval.
func Int8FromContext(ctx context.Context, key interface{}) (result Int8Result) {
	if v, ok := ctx.Value(key).(int8); !ok {
		result.Err(fmt.Errorf("%#v not found in context", key))
	} else {
		result.Ok(v)
	}
	return
}
